Add support for .docx files

gravesm opened the pull request:
This adds support for reading uploaded docx files. A document module is
added with a factory function for creating a document object from a
Django UploadedFile object. The created document object has a words
property which contains a list of words in the document. In addition,
this uses chardet to determine the encoding of plain text files.

thatandromeda commented on hamlet/theses/document.py:
```diff
@@ -0,0 +1,67 @@
+    """Document object for representing a DOCX document."""
+    def __init__(self, doc):
+        self._words = []
+        self.doc = docx.Document(doc)
```
I take it `docx` handles all encoding issues internally?

gravesm replied:
Yeah, by the time we get text out of it, it's already unicode. I will say that if we decide we don't like how the library I'm using (it's the only one I could find) is performing it probably wouldn't be totally onerous to roll our own. Our needs for pulling text out of the document are very minimal.

thatandromeda commented on hamlet/theses/fixtures/thesis.txt:
```diff
@@ -0,0 +1,3 @@
+Since the dawn of time, humans have been using hatsopoulos microfluids to accomplish some amazing things. Today, we stand upon the precipice of a new age where hatsopoulos microfluids will dictate all aspects of our lives.
```
🎉 😆

thatandromeda commented on hamlet/theses/forms.py:
```diff
@@ -85,16 +69,15 @@ class Meta:
-    allowed_filetypes = ['text/plain']
+    allowed_extensions = ['txt', 'docx']
+    allowed_mimetypes = ['text/plain',
+        'application/vnd.openxmlformats-officedocument.wordprocessingml.document']
```
Do you need a file generated from an actual Windows product to test this on? I can easily supply you with an OS X Office file, and I can get a Windows one also.

gravesm replied:
Probably wouldn't hurt, though I'd guess with what we're doing, any differences in implementation of the standard won't have much of an impact. It's no problem to throw a few more files at it during testing, so why not?

thatandromeda replied:
I've sent you a Mac-generated one, and Matt will be sending you a Windows-generated one shortly.

thatandromeda commented:
I fired this up and tried to upload a .docx and did not meet with success:

<img width="1316" alt="screen shot 2018-02-06 at 12 48 00 pm" src="https://user-images.githubusercontent.com/959761/35875012-12361328-0b3c-11e8-93d3-3650648bac94.png">
<img width="1327" alt="screen shot 2018-02-06 at 12 48 11 pm" src="https://user-images.githubusercontent.com/959761/35875013-1241f972-0b3c-11e8-9659-2a7c01c50fb6.png">

thatandromeda committed:
Specify version of gensim

gensim 3.3 has a backwards-incompatible change that broke our existing
file upload handling. I'll open a ticket to make this 3.3-compatible,
but for now this is the easy fix.

thatandromeda commented:
File upload wasn't working for me, but it turned out this was due to incompatibilities between gensim versions - I've pinned the version number in the Pipfile and now it works. I'll open a ticket for the version incompatibility issue (but it will be very low priority; rewriting the gensim code is thinky work and there are more important ways to spend our time).

thatandromeda reviewed (approved):
LGTM but tell me if you want to do anything with respect to the questions above.

gravesm commented:
I think I'd like to at least look into updating the tests to go through the actual view, based on the document you sent me before merging.

thatandromeda commented:
Integration tests ftw.

thatandromeda committed:
Update tests that work on local but fail on Travis

thatandromeda committed:
Add support for .docx files

This adds support for reading uploaded docx files. A document module is
added with a factory function for creating a document object from a
Django UploadedFile object. The created document object has a words
property which contains a list of words in the document. In addition,
this uses chardet to determine the encoding of plain text files.

thatandromeda committed:
Add test fixtures to allow for thesis testing

thatandromeda committed:
Add function tests for upload

This adds tests for the upload recommender view's ability to return
matching documents using both .txt and .docx uploads.

thatandromeda committed:
Use .env rather than personal settings to toggle model in use

gravesm commented on docs/developer.md:
```diff
@@ -79,3 +79,8 @@ hamlet.model is a copy of all_theses_no_split_w4_s52.model. This is a model trai
+`hamlet.settings.local` defaults to using the test model, since it is checked
+into version control. If you have a different model you want to use:
+* put it in `hamlet/model/hamlet.model` (along with any other numpy files it needs to work);
+* set `DJANGO_USE_LIVE_MODEL=True` in `.env`.
```
What if we added a new setting called something like HAMLET_MODEL that specified the path to an alternate model? This would have the advantage of being more flexible in where it lives and what it's called, as well as being more explicit about what's getting used. If the setting doesn't exist we can fall back to the testmodel.

thatandromeda replied:
Presumed to be the relative path from the project root? Sure, that works. Will do.
